### PR TITLE
Change support of ThermoPro TX-2C sensor to enable reading humidity data

### PR DESCRIPTION
**Pull request by mwerlen**
Hi,

I've added some more information on ThermoPro TX-2C when a humidity sensor is available.
Decoding pattern is similar to other ThermoPro devices.
There is working tests in rtl_433_tests in another pull request  (https://github.com/merbanan/rtl_433_tests/pull/453)

### CODE REVIEW

File: src/devices/thermopro_tx2c.c [hunk 1]
```diff
@@ -56,16 +70,29 @@ static int thermopro_tx2c_decode(r_device *decoder, bitbuffer_t *bitbuffer)
...
+        return DECODE_FAIL_SANITY;
+
+    int type     = b[0] >> 4;
+    int id       = (uint8_t)(((b[0] & 0xF) << 4) | (b[1] >> 4));
+    int battery  = (b[1] & 0x08) >> 3;
+    int button   = (b[1] & 0x04) >> 2;
+    int channel  = (b[1] & 0x03) + 1;
     int temp_raw = (int16_t)((b[2] << 8) | b[3]); // uses sign-extend
     float temp_c = (temp_raw >> 4) * 0.1f;
+    int humidity = (uint8_t)(((b[3] & 0xF) << 4) | (b[4] >> 4));
```
Comment (zuckschwerdt) [inline src/devices/thermopro_tx2c.c:84]: No need for a `(uint8_t)` cast.

**Review by zuckschwerdt — COMMENTED** [review]
Looks good, thanks. We need to think about `subtype` as that would be a breaking change.
Is it really a fixed model code or maybe a variable preamble?

File: src/devices/thermopro_tx2c.c [hunk 2]
```diff
@@ -26,17 +26,31 @@ Example data:
...
+  the same batteries often generate the same id
+- flags(3): is 1 when the battery is low, otherwise 0 (ok)
+- flags(2): is 1 when the sensor sends a reading when pressing the button on the sensor
+- flags(1,0): the channel number that can be set by the sensor (1, 2, 3, X)
+- temp: 12 bit signed scaled by 10
+- humi: 8 bit always 00001010 (0x0A) if no humidity sensor is available
+- zero : a trailing 12 bit fixed 000000000000
 
 */
+#define BYTE_TO_BINARY_PATTERN "%c%c%c%c%c%c%c%c"
```
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c:44]: Yes :)

File: src/devices/thermopro_tx2c.c [hunk 3]
```diff
@@ -1,7 +1,7 @@
 /** @file
-    ThermoPro TX-2C Outdoor Thermometer.
+    ThermoPro TX-2C Outdoor Thermometer and humidity sensor
```
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

File: src/devices/thermopro_tx2c.c [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
 
-    Copyright (C) 2023 igor@pele.tech.
+    Copyright (C) 2023 igor@pele.tech and maxime@werlen.fr
```
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

File: src/devices/thermopro_tx2c.c [hunk 5]
```diff
@@ -26,17 +26,31 @@ Example data:
     [07] {45} 95 00 ff e0 a0 00
     [08] {36} 95 00 ff e0 a0
 
-Data layout:
+The data is grouped in 9 nibbles
```
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

File: src/devices/thermopro_tx2c.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

File: src/devices/thermopro_tx2c.c (hunk 1 above)
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

File: src/devices/thermopro_tx2c.c [hunk 6]
```diff
@@ -56,16 +70,29 @@ static int thermopro_tx2c_decode(r_device *decoder, bitbuffer_t *bitbuffer)
         return DECODE_FAIL_SANITY;
     }
 
-    int device   = (b[0]);
+    // check existing 12 bit 0 trailer
+    if ((b[4] & 0x0F) != 0x00 || b[5] != 0x00)
+        return DECODE_FAIL_SANITY;
+
+    int type     = b[0] >> 4;
+    int id       = (uint8_t)(((b[0] & 0xF) << 4) | (b[1] >> 4));
```
[the remark this answers is not in the document]
Comment (mwerlen) [reply src/devices/thermopro_tx2c.c]: Changed!

**Comment by mwerlen** [comment]
I took subtype from another ThermoPro device : https://github.com/merbanan/rtl_433/blob/ac36025a84451d404f5cba828c1f9b666be33cca/src/devices/thermopro_tx2.c#L70
It's not useful to me. If you think it's better not to output it, I can change that.

All other requested changes have been resolved

**Comment by zuckschwerdt** [comment]
Just comment out (`//`) both `subtype` lines (80 and 95) for now, we'll merge this and solve that issue later.

**Comment by mwerlen** [comment]
PR updated to comment subtype output.
[PR](https://github.com/merbanan/rtl_433_tests/pull/453) on rtl_433_tests will be updated accordingly

**Pull request merged**